baseapp/grpcrouter.go

```diff
@@ -0,0 +1,112 @@
+package baseapp
+
+import (
+	gocontext "context"
+	"fmt"
+
+	gogogrpc "github.com/gogo/protobuf/grpc"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/encoding/proto"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var protoCodec = encoding.GetCodec(proto.Name)
+
+// GRPCRouter routes ABCI Query requests to GRPC handlers
+type GRPCRouter struct {
+	routes map[string]GRPCQueryHandler
+}
+
+var _ gogogrpc.Server
+
+// NewGRPCRouter creates a new GRPCRouter
+func NewGRPCRouter() *GRPCRouter {
+	return &GRPCRouter{
+		routes: map[string]GRPCQueryHandler{},
+	}
+}
+
+// GRPCQueryHandler defines a function type which handles ABCI Query requests
+// using gRPC
+type GRPCQueryHandler = func(ctx sdk.Context, req abci.RequestQuery) (abci.ResponseQuery, error)
+
+// Route returns the Querier for a given query route path.
+func (qrt *GRPCRouter) Route(path string) GRPCQueryHandler {
+	handler, found := qrt.routes[path]
+	if !found {
+		return nil
+	}
+	return handler
+}
+
+// RegisterService implements the grpc Server.RegisterService method
+func (qrt *GRPCRouter) RegisterService(sd *grpc.ServiceDesc, handler interface{}) {
+	// adds a top-level querier based on the GRPC service name
+	for _, method := range sd.Methods {
+		fqName := fmt.Sprintf("/%s/%s", sd.ServiceName, method.MethodName)
+		methodHandler := method.Handler
+
+		qrt.routes[fqName] = func(ctx sdk.Context, req abci.RequestQuery) (abci.ResponseQuery, error) {
+			res, err := methodHandler(handler, sdk.WrapSDKContext(ctx), func(i interface{}) error {
+				return protoCodec.Unmarshal(req.Data, i)
+			}, nil)
+			if err != nil {
+				return abci.ResponseQuery{}, err
+			}
+
+			resBytes, err := protoCodec.Marshal(res)
+			if err != nil {
+				return abci.ResponseQuery{}, err
+			}
+
+			return abci.ResponseQuery{
+				Height: req.Height,
+				Value:  resBytes,
+			}, nil
+		}
+	}
+}
+
+// QueryServiceTestHelper provides a helper for making grpc query service
```

Comment:
nit: move test types and logic to separate file.

Reply:
grpcrouter_helpers.go?